src/main/java/org/openjdk/shenandoah/ShenandoahVisualizer.java

```diff
@@ -22,79 +23,354 @@
+                        toolbarPanel.setLastActionField("Play button pressed.");
+                    } else {
+                        toolbarPanel.setLastActionField("Pause button pressed.");
+                    }
+                    renderRunner.playback.isPaused = !renderRunner.playback.isPaused;
+                }
+                if (e.getKeyCode() == KeyEvent.VK_UP) {
+                    renderRunner.playback.stepForwardSnapshots(5);
+                }
+                if (e.getKeyCode() == KeyEvent.VK_RIGHT) {
+                    renderRunner.playback.stepForwardSnapshots(1);
+                }
+                if (e.getKeyCode() == KeyEvent.VK_ENTER && lastSnapshotIndex > 0) {
+                    renderRunner.playback.stepForwardSnapshots(lastSnapshotIndex);
+                }
+            }
+        };
+        toolbarPanel.setPlayPauseButtonListener(playPauseButtonListener);
+        // Step back/forward button listeners
+        toolbarPanel.setBackButton_1_Listener((ae) -> renderRunner.playback.stepBackSnapshots(1));
+
+        toolbarPanel.setBackButton_5_Listener((ae) -> renderRunner.playback.stepBackSnapshots(5));
+
+        toolbarPanel.setForwardButton_1_Listener((ae) -> renderRunner.playback.stepForwardSnapshots(1));
+
+        toolbarPanel.setForwardButton_5_Listener((ae) -> renderRunner.playback.stepForwardSnapshots(5));
+
+        toolbarPanel.setEndSnapshotButtonListener(e -> {if (lastSnapshotIndex > 0) renderRunner.playback.stepForwardSnapshots(lastSnapshotIndex);});
+
+        ChangeListener sliderListener = new ChangeListener() {
+            @Override
+            public void stateChanged(ChangeEvent e) {
+                int difference = toolbarPanel.currentSliderValue() - renderRunner.playback.getPopupSnapshotsSize();
+                if (difference > 0) {
+                    renderRunner.playback.stepForwardSnapshots(difference);
+                }
+                if (difference < 0) {
+                    renderRunner.playback.stepBackSnapshots(Math.abs(difference));
+                }
+
             }
         };
+        toolbarPanel.setSliderListener(sliderListener);
 
+        // Speed button listeners
+        ChangeListener speedSpinnerListener = new ChangeListener() {
+            @Override
+            public void stateChanged(ChangeEvent e) {
+                if (!toolbarPanel.speedButtonPressed) {
+                    double speed = toolbarPanel.getSpeedValue();
+                    if (speed != renderRunner.playback.speed) {
+                        toolbarPanel.setLastActionField("Changed playback speed to: " + speed);
+                        renderRunner.playback.data.setSpeed(speed);
+                        renderRunner.playback.speed = speed;
+                    }
+                }
+            }
+        };
+        toolbarPanel.setSpeedSpinnerListener(speedSpinnerListener);
+
+        ActionListener speed_0_5_Listener = new ActionListener() {
```

Comment:
That seems an odd variable name (format). Maybe halfSpeedListener ?

Reply:
Agreed, I'll change this.